Mono created from a Sink.one sporadically turns empty

<!--- Provide a general summary of the issue in the Title above -->
In our code, we create a `Sink.one()` whose value is emitted from onSuccess/onError handlers of some other `Mono`. A value being emitted is never `null`.
<!--- /!\ Make sure to follow the Contribution Guidelines, notably for security issues and questions:
https://github.com/reactor/.github/blob/main/CONTRIBUTING.md
https://pivotal.io/security
https://github.com/reactor/.github/blob/main/CONTRIBUTING.md#question-do-you-have-a-question
-->

## Expected Behavior
Expectation is that a mono view of that sink will never be an empty mono, due to value emitted by a sink being non-null.
<!--- Tell us what you think should happen. -->

## Actual Behavior
Contary to our expectation, a Mono view of that sink _sporadically_ turns empty. It happens both on Windows 10 as well as on Ubuntu 20 machines, so it is likely not something platform-specific.

## Steps to Reproduce
Below is the shortest unit test I could come up with which reflects what we're doing in our production code, and when this test is run from IDE with "repeat until failure" setting, it fails the assertion about non-null Mono result quite consistently within 10-30 seconds after start.

I also added some diagnostic logging around tryEmitValue / tryEmitError, and looks like it's never an error, and never a null value being emitted, neither it's an emission failure.

```java
import java.util.concurrent.*;
import java.util.function.Consumer;

import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.*;
import reactor.core.scheduler.Schedulers;
import reactor.util.Loggers;

import static org.assertj.core.api.AssertionsForClassTypes.assertThat;

@Slf4j
class MonoPlaygroundTest {

    private static ExecutorService executor;

    static {
        executor = Executors.newCachedThreadPool();
        Loggers.useSl4jLoggers();
    }

    @Test
    void shouldConstructNonEmptyMono() {
        var sink = Sinks.<String>one();

        Mono
            .fromSupplier(() -> "value")
            .log(Loggers.getLogger("sink-setter"))
            .subscribeOn(Schedulers.fromExecutorService(executor, "mono-executor"))
            .doOnSuccess(tryEmitValueWithDebugLogging(sink))
            .doOnError(tryEmitErrorWithDebugLogging(sink))
            .subscribe();

        var result = sink
            .asMono()
            .log(Loggers.getLogger("sink-result-receiver"))
            // more map / flatMap here...
            .subscribeOn(Schedulers.boundedElastic())
            .block();

        assertThat(result).isNotNull();
    }

    private static Consumer<Throwable> tryEmitErrorWithDebugLogging(Sinks.One<String> sink) {
        return t -> {
            sink.tryEmitError(t);
            log.error("Was emitting an error: {}", t);
        };
    }

    private static Consumer<String> tryEmitValueWithDebugLogging(Sinks.One<String> sink) {
        return value -> {
            var result = sink.tryEmitValue(value);
            if (value == null) {
                log.error("Tried to emit null value");
            }
            if (result.isFailure()) {
                log.error("Failed to emit because of {}", result);
            }
        };
    }
}
```

The output from a failed run is like this:
```
23:48:12.635 [pool-1-thread-1] INFO  sink-setter - | onSubscribe([Fuseable] Operators.MonoSubscriber)
23:48:12.635 [pool-1-thread-1] INFO  sink-setter - | request(unbounded)
23:48:12.635 [boundedElastic-1] INFO  sink-result-receiver - onSubscribe(SinkOneMulticast.NextInner)
23:48:12.635 [pool-1-thread-1] INFO  sink-setter - | onNext(value)
23:48:12.635 [boundedElastic-1] INFO  sink-result-receiver - request(unbounded)
23:48:12.635 [pool-1-thread-1] INFO  sink-setter - | onComplete()
23:48:12.635 [boundedElastic-1] INFO  sink-result-receiver - onComplete()

java.lang.AssertionError: 
Expecting actual not to be null
```

## Possible Solution
<!--- Not obligatory, but you can suggest a fix/reason for the bug. -->

## Your Environment

<!--- Include as many relevant details about the environment you experienced the bug in. -->
<!--- Especially, always include the version(s) of Reactor library/libraries you used! -->

* Reactor version(s) used: 3.4.22
* JVM version (`java -version`): openjdk version "11.0.16.1" 2022-07-19 LTS
* OS and version (eg `uname -a`): Windows 10 / 21H2 build 19044.2130, also Linux 5.4.0-96-generic SMP x86_64 GNU/Linux

--- a/reactor-core/src/jcstress/java/reactor/core/publisher/SinkOneStressTest.java
+++ b/reactor-core/src/jcstress/java/reactor/core/publisher/SinkOneStressTest.java
@@ -1,0 +1,112 @@
+/*
+ * Copyright (c) 2020-2022 VMware Inc. or its affiliates, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.core.publisher;
+
+import org.openjdk.jcstress.annotations.Actor;
+import org.openjdk.jcstress.annotations.Arbiter;
+import org.openjdk.jcstress.annotations.JCStressTest;
+import org.openjdk.jcstress.annotations.Outcome;
+import org.openjdk.jcstress.annotations.State;
+import org.openjdk.jcstress.infra.results.LLL_Result;
+import org.openjdk.jcstress.infra.results.LL_Result;
+
+import static org.openjdk.jcstress.annotations.Expect.ACCEPTABLE;
+
+public class SinkOneStressTest {
+
+	@JCStressTest
+	@Outcome(id = {"foo, 1"}, expect = ACCEPTABLE, desc = "value delivered")
+	@State
+	public static class EmitValueSubscribeStressTest extends SinkOneStressTest {
+
+		final SinkOneMulticast<String> sink = new SinkOneMulticast<>();
+		final StressSubscriber<String> subscriber = new StressSubscriber<>();
+
+		@Actor
+		public void value() {
+			sink.tryEmitValue("foo");
+			sink.tryEmitEmpty();
+			sink.tryEmitError(new RuntimeException());
+		}
+
+		@Actor
+		public void subscriber() {
+			sink.asMono().subscribe(subscriber);
+		}
+
+		@Arbiter
+		public void arbiter(LL_Result r) {
+			r.r1 = subscriber.onNextCalls.get() == 1 ? subscriber.receivedValues.get(0) : null;
+			r.r2 = subscriber.onCompleteCalls.get();
+		}
+	}
+
+	@JCStressTest
+	@Outcome(id = {"null, 1"}, expect = ACCEPTABLE, desc = "value delivered")
+	@State
+	public static class EmitEmptySubscribeStressTest extends SinkOneStressTest {
+
+		final SinkOneMulticast<String> sink = new SinkOneMulticast<>();
+		final StressSubscriber<String> subscriber = new StressSubscriber<>();
+
+		@Actor
+		public void value() {
+			sink.tryEmitEmpty();
+			sink.tryEmitValue("foo");
+			sink.tryEmitError(new RuntimeException());
+		}
+
+		@Actor
+		public void subscriber() {
+			sink.asMono().subscribe(subscriber);
+		}
+
+		@Arbiter
+		public void arbiter(LL_Result r) {
+			r.r1 = subscriber.onNextCalls.get() == 1 ? subscriber.receivedValues.get(0) : null;
+			r.r2 = subscriber.onCompleteCalls.get();
+		}
+	}
+
+	@JCStressTest
+	@Outcome(id = {"null, 0, boo"}, expect = ACCEPTABLE, desc = "value delivered")
+	@State
+	public static class EmitErrorSubscribeStressTest extends SinkOneStressTest {
+
+		final SinkOneMulticast<String> sink = new SinkOneMulticast<>();
+		final StressSubscriber<String> subscriber = new StressSubscriber<>();
+
+		@Actor
+		public void value() {
+			sink.tryEmitError(new RuntimeException("boo"));
+			sink.tryEmitEmpty();
+			sink.tryEmitValue("foo");
+		}
+
+		@Actor
+		public void subscriber() {
+			sink.asMono().subscribe(subscriber);
+		}
+
+		@Arbiter
+		public void arbiter(LLL_Result r) {
+			r.r1 = subscriber.onNextCalls.get() == 1 ? subscriber.receivedValues.get(0) : null;
+			r.r2 = subscriber.onCompleteCalls.get();
+			r.r3 = subscriber.onErrorCalls.get() == 1 ? subscriber.error.getMessage() : null;
+		}
+	}
+}

--- a/reactor-core/src/main/java/reactor/core/publisher/SinkEmptyMulticast.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/SinkEmptyMulticast.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020-2021 VMware Inc. or its affiliates, All Rights Reserved.
+ * Copyright (c) 2020-2022 VMware Inc. or its affiliates, All Rights Reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -37,9 +37,14 @@ class SinkEmptyMulticast<T> extends Mono<T> implements InternalEmptySink<T> {
 
 	@SuppressWarnings("rawtypes")
 	static final Inner[] EMPTY = new Inner[0];
-
 	@SuppressWarnings("rawtypes")
-	static final Inner[] TERMINATED = new Inner[0];
+	static final Inner[] TERMINATED_EMPTY = new Inner[0];
+	@SuppressWarnings("rawtypes")
+	static final Inner[] TERMINATED_ERROR = new Inner[0];
+
+	static final int STATE_ADDED = 0;
+	static final int STATE_ERROR = -1;
+	static final int STATE_EMPTY = -2;
 
 	@Nullable
 	Throwable error;
@@ -58,12 +63,23 @@ class SinkEmptyMulticast<T> extends Mono<T> implements InternalEmptySink<T> {
 		return this;
 	}
 
+	boolean isTerminated(Inner<?>[] array) {
+		return array == TERMINATED_EMPTY || array == TERMINATED_ERROR;
+	}
+
 	@Override
 	public EmitResult tryEmitEmpty() {
-		Inner<?>[] array = SUBSCRIBERS.getAndSet(this, TERMINATED);
+		Inner<T>[] array;
+		for (;;) {
+			array = this.subscribers;
 
-		if (array == TERMINATED) {
-			return Sinks.EmitResult.FAIL_TERMINATED;
+			if (isTerminated(array)) {
+				return EmitResult.FAIL_TERMINATED;
+			}
+
+			if (SUBSCRIBERS.compareAndSet(this, array, TERMINATED_EMPTY)) {
+				break;
+			}
 		}
 
 		for (Inner<?> as : array) {
@@ -77,23 +93,36 @@ class SinkEmptyMulticast<T> extends Mono<T> implements InternalEmptySink<T> {
 	public EmitResult tryEmitError(Throwable cause) {
 		Objects.requireNonNull(cause, "onError cannot be null");
 
-		Inner<T>[] prevSubscribers = SUBSCRIBERS.getAndSet(this, TERMINATED);
-		if (prevSubscribers == TERMINATED) {
+		Inner<T>[] prevSubscribers = this.subscribers;
+
+		if (isTerminated(prevSubscribers)) {
 			return EmitResult.FAIL_TERMINATED;
 		}
 
 		error = cause;
 
+		for (;;) {
+			if (SUBSCRIBERS.compareAndSet(this, prevSubscribers, TERMINATED_ERROR)) {
+				break;
+			}
+
+			prevSubscribers = this.subscribers;
+			if (isTerminated(prevSubscribers)) {
+				return EmitResult.FAIL_TERMINATED;
+			}
+		}
+
 		for (Inner<T> as : prevSubscribers) {
 			as.error(cause);
 		}
+
 		return EmitResult.OK;
 	}
 
 	@Override
 	public Object scanUnsafe(Attr key) {
-		if (key == Attr.TERMINATED) return subscribers == TERMINATED;
-		if (key == Attr.ERROR) return error;
+		if (key == Attr.TERMINATED) return isTerminated(subscribers);
+		if (key == Attr.ERROR) return subscribers == TERMINATED_ERROR ? error : null;
 		if (key == Attr.RUN_STYLE) return Attr.RunStyle.SYNC;
 
 		return null;
@@ -104,12 +133,16 @@ class SinkEmptyMulticast<T> extends Mono<T> implements InternalEmptySink<T> {
 		return Operators.multiSubscribersContext(subscribers);
 	}
 
-	boolean add(Inner<T> ps) {
+	int add(Inner<T> ps) {
 		for (; ; ) {
 			Inner<T>[] a = subscribers;
 
-			if (a == TERMINATED) {
-				return false;
+			if (a == TERMINATED_EMPTY) {
+				return STATE_EMPTY;
+			}
+
+			if (a == TERMINATED_ERROR) {
+				return STATE_ERROR;
 			}
 
 			int n = a.length;
@@ -118,7 +151,7 @@ class SinkEmptyMulticast<T> extends Mono<T> implements InternalEmptySink<T> {
 			b[n] = ps;
 
 			if (SUBSCRIBERS.compareAndSet(this, a, b)) {
-				return true;
+				return STATE_ADDED;
 			}
 		}
 	}
@@ -165,20 +198,21 @@ class SinkEmptyMulticast<T> extends Mono<T> implements InternalEmptySink<T> {
 	public void subscribe(final CoreSubscriber<? super T> actual) {
 		Inner<T> as = new VoidInner<>(actual, this);
 		actual.onSubscribe(as);
-		if (add(as)) {
+		final int addedState = add(as);
+		if (addedState == STATE_ADDED) {
 			if (as.isCancelled()) {
 				remove(as);
 			}
 		}
-		else {
+		else if (addedState == STATE_ERROR) {
 			Throwable ex = error;
-			if (ex != null) {
-				actual.onError(ex);
-			}
-			else {
-				as.complete();
-			}
+
+			actual.onError(ex);
 		}
+		else {
+			as.complete();
+		}
+
 	}
 
 	@Override

--- a/reactor-core/src/main/java/reactor/core/publisher/SinkOneMulticast.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/SinkOneMulticast.java
@@ -25,58 +25,51 @@ import reactor.util.annotation.Nullable;
 
 final class SinkOneMulticast<O> extends SinkEmptyMulticast<O> implements InternalOneSink<O> {
 
+	@SuppressWarnings("rawtypes")
+	static final Inner[] TERMINATED_VALUE = new Inner[0];
+
+	static final int STATE_VALUE = 1;
+
 	@Nullable
 	O value;
 
 	@Override
-	public EmitResult tryEmitEmpty() {
-		return tryEmitValue(null);
-	}
-
-	@Override
-	@SuppressWarnings("unchecked")
-	public EmitResult tryEmitError(Throwable cause) {
-		Objects.requireNonNull(cause, "onError cannot be null");
-
-		Inner<O>[] prevSubscribers = SUBSCRIBERS.getAndSet(this, TERMINATED);
-		if (prevSubscribers == TERMINATED) {
-			return EmitResult.FAIL_TERMINATED;
-		}
-
-		error = cause;
-		value = null;
-
-		for (Inner<O> as : prevSubscribers) {
-			as.error(cause);
-		}
-		return EmitResult.OK;
+	boolean isTerminated(Inner<?>[] array) {
+		return array == TERMINATED_VALUE || super.isTerminated(array);
 	}
 
 	@Override
 	public EmitResult tryEmitValue(@Nullable O value) {
-		@SuppressWarnings("unchecked") Inner<O>[] array = SUBSCRIBERS.getAndSet(this, TERMINATED);
-		if (array == TERMINATED) {
+		Inner<O>[] prevSubscribers = this.subscribers;
+
+		if (isTerminated(prevSubscribers)) {
 			return EmitResult.FAIL_TERMINATED;
 		}
 
 		this.value = value;
-		if (value == null) {
-			for (Inner<O> as : array) {
-				as.complete();
+
+		for (;;) {
+			if (SUBSCRIBERS.compareAndSet(this, prevSubscribers, TERMINATED_VALUE)) {
+				break;
+			}
+
+			prevSubscribers = this.subscribers;
+			if (isTerminated(prevSubscribers)) {
+				return EmitResult.FAIL_TERMINATED;
 			}
 		}
-		else {
-			for (Inner<O> as : array) {
-				as.complete(value);
-			}
+
+		for (Inner<O> as : prevSubscribers) {
+			as.complete(value);
 		}
+
 		return EmitResult.OK;
 	}
 
 	@Override
 	public Object scanUnsafe(Attr key) {
-		if (key == Attr.TERMINATED) return subscribers == TERMINATED;
-		if (key == Attr.ERROR) return error;
+		if (key == Attr.TERMINATED) return isTerminated(subscribers);
+		if (key == Attr.ERROR) return subscribers == TERMINATED_ERROR ? error : null;
 		if (key == Attr.RUN_STYLE) return Attr.RunStyle.SYNC;
 
 		return null;
@@ -86,24 +79,47 @@ final class SinkOneMulticast<O> extends SinkEmptyMulticast<O> implements Interna
 	public void subscribe(final CoreSubscriber<? super O> actual) {
 		NextInner<O> as = new NextInner<>(actual, this);
 		actual.onSubscribe(as);
-		if (add(as)) {
+		final int addState = add(as);
+		if (addState == STATE_ADDED) {
 			if (as.isCancelled()) {
 				remove(as);
 			}
 		}
+		else if (addState == STATE_ERROR) {
+			actual.onError(error);
+		}
+		else if (addState == STATE_EMPTY) {
+			as.complete();
+		}
 		else {
-			Throwable ex = error;
-			if (ex != null) {
-				actual.onError(ex);
+			as.complete(value);
+		}
+	}
+
+	@Override
+	int add(Inner<O> ps) {
+		for (; ; ) {
+			Inner<O>[] a = subscribers;
+
+			if (a == TERMINATED_EMPTY) {
+				return STATE_EMPTY;
 			}
-			else {
-				O v = value;
-				if (v != null) {
-					as.complete(v);
-				}
-				else {
-					as.complete();
-				}
+
+			if (a == TERMINATED_ERROR) {
+				return STATE_ERROR;
+			}
+
+			if (a == TERMINATED_VALUE) {
+				return STATE_VALUE;
+			}
+
+			int n = a.length;
+			@SuppressWarnings("unchecked") Inner<O>[] b = new Inner[n + 1];
+			System.arraycopy(a, 0, b, 0, n);
+			b[n] = ps;
+
+			if (SUBSCRIBERS.compareAndSet(this, a, b)) {
+				return STATE_ADDED;
 			}
 		}
 	}

